### ISSUE TITLE
In H2H combat, calculations should always be your attack vs their defense and vice versa

### ISSUE THREAD

**Issue by Amazinite** [endless-sky/endless-sky#9777]
### Problem Description

The current boarding mechanics are as follows:
* The player chooses whether to attack or defend.
* The AI chooses whether to attack or defend.
  * Whether the AI attacks or defense is based on some ratio between their attack stat and your defense stat.
* If both defend, combat ends.
* When you attack, your attack stat is rolled against the stat for whichever action the AI chose. That is, if the AI chose to attack then it is your attack vs their attack, while if the AI chose to defend then it is your attack vs their attack.

Given that defense stats are always higher than attack stats on ships, this creates an incentive for the player to lower their defense as much as possible while keeping their attack power high in order to roll your attack power against the AI's attack power, which is the weaker of the two stats.

### Related Issue Links

#6986
#2392

### Desired Solution

H2H combat should always be your attack vs their defense and their attack vs your defense. Whether the AI chooses to attack or defend should only influence whether you are able to flee combat. This way, the player actually has to worry about their defense stat and is actually incentivized to increase it.

This will also make capturing ships more difficult, as you will now always be going up against the ship's higher defense stat rather than the current scenario where you may roll against their easier attack stat for a few actions before the AI decides to start defending.

### Alternative Approaches

N/A

**Comment by Zitchas** [comment]
As an aside, I fundamentally don't think we should be running statistical analysis of likely combat results at all. I don't even think we should be giving total attack and defense values. The player should know that for themselves, sure. But not their target. Likewise the AI shouldn't really know that about the player, either. 

H2H combat shouldn't be min-maxing a percentage win chance; it should be about prepping your ship as best you can, then hitting that attack button and seeing how things go. If you do well.... Maybe you were lucky, maybe you are better equipped than your target. If you don't do well, maybe you were unlucky, maybe you are underequipped. Over mutliple turns people will notice a trend and decide appropriate. People should need to think about it, not just glance at percentages and hammer the appropriate button. 

That tangent aside, yes, doing something to ensure the AI isn't going to suicide into the teeth of an attack would be good.

Actually, I think my preferred solution would probably be to take the player's actions into account. How often has the player hit attack in this boarding action? If they attacked last round, then assume they are going to attack this round and compare against their attack value instead of their defend value to determine if attacking or defending. Even if it's just that simple binary check, that'd eliminate a lot of the abuse of the mechanic. Or at least make it more challenging. It might still be possible to trick it in some situations - which is fine. The point isn't to utterly remove the ability to trick the AI ever; but rather to make it something that can be done now and then, but is hard or even impossible to do reliably and consistently in any particular boarding action; and even better if it is costly when it fails. 

For instance, going up against a big ship, if the player manages to trick it into attacking vs their attack in one or two turns... That's not going to be very significant when the overall battle can run 20-30+ turns. It might make the difference now and then, though, and that's OK. Good, even. It makes H2H combat require more skill, and be less predictable. Both good things.

**Comment by Quantumshark** [comment]
I'm not entirely sure I understand how this would work.
Right now, we have a system where both sides choose to attack (i.e. use their attack), or defend (i.e. use their defence), with a specific scenario where if both sides defend no combat takes place.
So, how exactly do you imagine this working?
As far as I can tell, your desired solution is to run two 'rounds' of combat each turn, one with the player's attack against the npc's defense, and one with the npc's attack against the player's defense, and keep the both defend = no combat thing. But, that would mean the attack/defend choice makes no difference unless you're trying to end combat, which is maybe not ideal?

Also, side note:
> We should massively buff the security station's defense.

No. We should massively _nerf_ the defense of the three h2h outfits with stats an order of magnitude higher than anything else - the Microbot Defense Station (45), the Intrusion Cuntermeasures (60) and the Pug Biodefenses (250). Right now, we _can't_ make defense useful for the player without trivialising capturing, because these three outfits, once you have a source of any of them, make it trivial to get absurdly high defense stats that guarantee victories in any engagement the player's defense is involved in.

**Comment by Quantumshark** [comment]
> Likewise the AI shouldn't really know that about the player, either.

This is an important point. I don't think the AI should be able to act on any information that we opt to deny the player access to. Right now, both the AI and the player have perfect knowledge of the odds, which a) doesn't make much sense in-universe, b) is very computationally expensive, and c) isn't really great for gameplay, either. But, if we want to take information away from the player, then to avoid giving the AI an advantage which is unfair, and still just as nonsensical and computationally expensive as before, then we need to be open to the fact that we'll have boarding AI that sometimes makes the wrong choice.

**Comment by alexrovw** [comment]
I think this makes perfect sense.

In the current game state, the player's defense stat is useless. In fact, it's worse than useless, because of the exploitative strategy of minimizing your own defense (typically via Nerve Gas) in order to bait the AI into attacking to its own detriment. The AI is being forced to make a stupid decision because of the way that our existing boarding mechanics work.

In the proposed solution, both your attack and defense stats would be vital. Your attack stat would clash against the opponent's defense stat to determine how many casualties the opponent suffers. Your defense stat would clash against the opponent's attack stat to determine how many casualties _you_ suffer.

This is a great idea. Defensive outfits would no longer be useless clutter for the player, as they would protect the lives of your crew. We would basically be changing the "Defend" button into an "Attempt to Flee" button. You could only flee in the situation where the enemy is on the defensive. I like that. The only change I would make is to introduce a chance of successfully fleeing _for a price_ when you realize you're over your head. Maybe half of your surviving crew gets left behind if you forcibly flee from a ship while the enemy is on the attack. It's kind of weird to imagine that you wouldn't be able to escape from a disabled ship during a boarding action gone awry.

**Comment by Zitchas** [comment]
That sounds good. So instead of having "attack" and "defend", the two options would be "press the engagement" and "refrain from engaging". 

Off course this also ties into the whole fact that there player is never meaningfully boarded.

**Comment by alexrovw** [comment]
> Of course this also ties into the whole fact that there player is never meaningfully boarded.

Yeah, true. I kind of miss the mechanic in EVN where you lost some credits or cargo when you were boarded. We can't take advantage of that fully, though, because we don't have an "iron man" or "hardcore" mode. The player will reload their save instead of suffering the consequences of being boarded.

If we were to have something like that, I would expand Bank mechanics to be able to "save" some of your credits in your bank account to protect them from boarders. But I don't want to go too far off topic, so I'll stop there. 😅

**Comment by Amazinite** [comment]
> So, how exactly do you imagine this working?
As far as I can tell, your desired solution is to run two 'rounds' of combat each turn, one with the player's attack against the npc's defense, and one with the npc's attack against the player's defense

Think of the stats like attack and defense stats in an RPG. Attack increases how much damage you deal to the opponent while defense reduces how much damage you take from the opponent.

Alex and Zitchas correctly point out that this means that the meaning of the "Attack" and "Defend" buttons changes, and perhaps the buttons could be renamed. 

I do recognize that this doesn't make perfect sense with some of our current H2H weapons; how does a security station on your ship prevent you from dying when you're pressing the engagement into the enemy's ship? I'd rather ignore that for now, as the nature of our existing H2H weapons shouldn't dictate how the mechanics of attack and defense function, as H2H outfits are nothing but their attack and defense stats.

**Comment by xX-Dillinger-Xx** [comment]
As a player I see it this way. 
Ignoring stats and calculations for the moment and viewing it from the perspective of our character. If I'm boarding, I'm attacking. If I fail the attack, I'm retreating and better have good enough defenses to avoid being boarded myself. If I fail to defend my ship, that's on me. I should know if I have a strong crew or not. Not relative to the ship I'm attacking, but relative to an average of similarly sized crews on a similar class of ship. I'm not sure what the consequences of a failed attempt at capture should be, but it kinda doesn't matter because everyone will just hit the _magic go back in time_ button. (ESC)

I should be able to know what their crew compliment is. That information would be available upon disabling and docking with said ship. What I wouldn't know is how well they are armed, what defenses they have, and how well trained there combat and defence forces are. Also it should depend on the race you are attacking. As an example, 1 highly trained Heliarc should be equal to 2 humans.
 
On a side note that's a bit off topic but related, the whole capturing of ships seems oddly simplistic as it is. How, after I just shot the crap out of a ship enough to disable it, am I just able to board it, disable the crew, replace the crew, don't do any repairs, and magically the replacement crew knows everything they need to know about the alien technology  to just fly away? And what happened to the remainder of their crew? Are they now prisoners, did some choose to join me, or are they all dead?
And that leads me to another side note. How am I able hire new crew from an alien planet that have never seen humans before? Yes, I know that different alien races can be crew members of the same crew, but it just seems there should be more restrictions on where we can get replacement crew. 

IMO it would be perfectly acceptable, maybe even preferred, to expand the whole boarding, looting, and capturing scenarios into something much more immersive and drawn out. How to accomplish all this is up to the code crunchers.

**Comment by Zitchas** [comment]
Regarding target ship info, why should we have precise counts as soon as we board? If someone wants precise counts, they should need to use a tool to get it. Tactical scanner for crew count, outfit scan for outfits. Both combined could give an approximate strength, although still wouldn't know the strength of a crew member.

**Comment by xX-Dillinger-Xx** [comment]
Currently, even without scanners, I'm given the crew count and their strength stats the moment I 'dock'. That doesn't make sense.  So yes, having to have various scanners installed should be required to have access to that information.
 Another thing to think about, should "required crew" count towards your attack strength? I could see loosing some during the battle, but they shouldn't be directly involved in the attack. No, captain is going to send **all** the crew to attack. However, they would become relevant to the defence.

**Comment by Quotatron3000** [comment]
> if we want to take information away from the player

For me, the issue with this is attack/defence values can wildly vary from ship to ship, even if there are similar crew numbers. For example, a Korath ship can have an attack of +800, and a defence of +1000 for 140 to 250 troops, meanwhile your own ship may have only 200 to 300 for either stat in the early game. If you attempt to board a Korath ship without knowing these figures, it guarantees your defeat, every single time.

It isn't just 'oh, I lost my boarding party' but 'you lost your entire ship, game over, reload'. Any situation that involves a player losing, due to seemingly random, unknown numbers, is a frustrating experience (it makes it extremely difficult to 'learn from mistakes'). And the end result is they will either never board out of frustration, or just look up ballpark figures in a Wiki entry somewhere anyway which makes the experience overall less fun.

People argue this information shouldn't be known, but we're in a universe with outfit, tactical, asteroid and value scanners that can tell you the entire contents of a ship. Knowing these values is both in-universe, and required to allow the player to make sensible decisions without feeling like they're getting 'bad RNG'.

**Comment by Hurleveur** [comment]
You shouldnt expect to be able to capture a worldship with a lowly 200 crew, and also the info could be present if you use a scanner of some kind.
Deterring players from boarding too much would be a good thing, I think.

**Comment by Quotatron3000** [comment]
> You shouldnt expect to be able to capture a worldship with a lowly 200 crew

It isn't self-evident within the game that it is a 'worldship' (most warships appear as very similar sizes). It won't be a thing a player will know immediately. And I wasn't suggesting one should expect to - I'm saying the information is useful for informing a player it isn't meaningfully possible. I don't need to know it's a worldship if it has 800 attack.

>Deterring players from boarding too much would be a good thing

By forcing numerous reloads with saves? Sounds very frustrating. Boarding of powerful ships is already discouraged with the high attack figures (which is fine). What I'm saying is, if boarding values are too unknown, *newer players won't even try to board* - making it an unused feature. And experienced players will just Wiki the values, making hiding of the values redundant.

> also the info could be present if you use a scanner of some kind.
Tactical scanners already offer person counts without boarding, as does the Unfettered 'value scanner' without boarding (in-fact, the value scanner's sole purpose is for determining the value of items before boarding). That said, even without the scanners, many descriptions for various planets and ports refer to 'lifeforms' (including descriptors and quantities); one would presume the future has a lifeform scanner by default on ships.

The only situation I could see it being an unknown quantity is if interference plating is used. But my thoughts are this doesn't address the original issue's point of attack/defence values being 'sort of' redundant (hiding them would not address their point).

Anyway I'm supposed to be hiding. : )

**Comment by Hurleveur** [comment]
~~If you're boarding something 5 times your size you should expect a big crew~~
Also yeah the information would just stick in battle, with additional def and attack info but no odds, and without a scanner you're left with estimation, maybe? Like relative to your current crew we could write "overwhelming crew" or in the odds write smth bad

**Comment by brendanjones** [comment]
I’d like to suggest keeping the chat here to resolving the attack/defence mechanic, which IMO can be resolved without making any other changes like to stat visibility. I do like the discussion though, it’s interesting and worth having :)

**Comment by Hurleveur** [comment]
The probabilities are no longer accurate with this new system, so they either have to be adapted or removed.

**Comment by xX-Dillinger-Xx** [comment]
If I understand it correctly, calculating the **ODDS** requires too much overhead. If that is correct, I vote we scrap the **ODDS** and just stick with bare static stats for crew counts, attack strength and defensive strength. I do find it odd that either side would know how strong the others crew is, but can except that we need to know that in order to decide whether to attack or not. I suppose one could believe the the info comes from installed scanners. If that's the case we should expect some inaccuracy, therefore there is still a random aspect to the attack.

**Comment by Zitchas** [comment]
> I suppose one could believe the the info comes from installed scanners

The problem is that for many people, especially those pushing the edges to capture the biggest thing they have access to, scanners take up valuable space, and thus aren't used. Actually, I get the impression that a lot of people don't install scanners at all unless they really need to for a mission. I personally always like to have scanners and spend the space even when not optimal to do so, but I suspect I am in the minority.

**Comment by brendanjones** [comment]
> In any case, I'm 100% in favor of deleting the odds from the display. Leaving the attack strength is OK for the purposes of this particular issue, but I'd like to see the enemy ship's strength and crew count changed to only visible when the player has scanners capable of revealing crew count and outfits respectively as a long-term change here.

If this is done it’d be good to introduce an early-game mission that has the player install and make use of scanning for crew, to teach that it’s possible to reveal that stat. It’s pretty crucial info to have when you’re capping ships - not just for knowing that you can take the ship but also for estimating how much of your crew will be needed to pilot it once you succeed in capturing it.

**Comment by Hyratel** [comment]
quick brainstorm for a mission:
Pre-Condition: Player has no Marauder ships in their fleet (in-flight or parked)
Single ship: Marauder Firebird (chosen for high-in-class bunk count)
Ship is Faction: Merchant
"Pirate Marauder has been spotted around %system%. Scan this ship (without attacking) so system defense can esimate the threat"
Requires: Scanner with Tactical 
Attacking Marauder fails the mission and gets System Authority (temporarily) Mad At You. 
Scan reveals: maxed crew (29: 25+4 from 1 Bunkroom)
Popup message: "That ship is outfitted for piracy and boarding! Land and make sure you're equipped to handle that."
-Land- 
Popup message explaining boarding, looting, capturing mechanics, making the point about Continuous (beam) weapons being more suitable because they don't have Stop-Firing Lag when the target is disabled
-Launch- 
Target Marauder: Marauder turns hostile, player attacks. 
Marauder disabled: Popup from System Authority, "Good shooting! the ship is now able to be boarded. Bring it in so we can investigate their equipment." Mission will not spawn again.
Marauder destroyed: Popup. "Damn! I wanted a chance to look over what they had." Mission can spawn again.

**Comment by Quantumshark** [comment]
Closing as resolved by #10060.
